warehouse/jobs/runner.go

```diff
@@ -0,0 +1,373 @@
+	VALUES ($1, $2, $3, $4, $5, $6 ,$7, $8, $9, $10,$11,$12) RETURNING id`, warehouseutils.WarehouseAsyncJobTable)
+
+	stmt, err := asyncWhJob.dbHandle.Prepare(sqlStatement)
+	if err != nil {
+		asyncWhJob.log.Errorf("Error preparing out the query %s ", sqlStatement)
+		err = errors.New("error preparing out the query, while addJobstoDb")
+		return
+	}
+
+	defer stmt.Close()
+	now := timeutil.Now()
+	row := stmt.QueryRow(payload.SourceID, payload.Namespace, payload.DestinationID, payload.DestType, payload.JobRunID, payload.TaskRunID, payload.TableName, WhJobWaiting, now, now, payload.JobType, payload.AsyncJobType)
+	err = row.Scan(&jobId)
+	if err != nil {
+		asyncWhJob.log.Errorf("Error processing the %s ", sqlStatement)
+		err = errors.New("error processing the query, while addJobstoDb")
+		return
+	}
+	return
+}
+
+/*
+Async Job runner's main job is to
+1) Scan the database for entries into wh_async_jobs
+2) Publish data to pg_notifier queue
+3) Move any executing jobs to waiting
+*/
+func (asyncWhJob *AsyncJobWhT) InitAsyncJobRunner() {
+	//Start the asyncJobRunner
+	asyncWhJob.log.Info("WH-Jobs: Initializing async job runner")
+
+	for retry := 0; retry < MaxCleanUpRetries; retry++ {
+		err := AsyncJobWH.cleanUpAsyncTable()
+		if err == nil {
+			AsyncJobWH.enabled = true
+			break
+		}
+		asyncWhJob.log.Infof("WH-Jobs: Cleanup asynctable failed with error %s", err.Error())
+	}
+
+	if asyncWhJob.enabled {
+		ctx, cancel := context.WithCancel(context.Background())
+		asyncWhJob.Cancel = cancel
+		g, ctx := errgroup.WithContext(ctx)
+		//How does threading policy work here? Should we have one thread per customer/namespace?
+		g.Go(func() error {
+			asyncWhJob.startAsyncJobRunner(ctx)
+			return nil
+		})
+		g.Wait()
+	}
+
+}
+
+func (asyncWhJob *AsyncJobWhT) cleanUpAsyncTable() error {
+	asyncWhJob.log.Info("WH-Jobs: Cleaning up the zombie asyncjobs")
+	sqlStatement := fmt.Sprintf(`UPDATE %s SET status='%s' WHERE status='%s'`, warehouseutils.WarehouseAsyncJobTable, WhJobWaiting, WhJobExecuting)
+	asyncWhJob.log.Infof("WH-Jobs: resetting up async jobs table query %s", sqlStatement)
+	_, err := asyncWhJob.dbHandle.Query(sqlStatement)
+	return err
+}
+
+/*
+startAsyncJobRunner is the main runner that
+1) Periodically queries the db for any pending async jobs
+2) Groups them together
+3) Publishes them to the pgnotifier
+4) Spawns a subroutine that periodically checks for responses from pgNotifier/slave worker post trackBatch
+*/
+func (asyncWhJob *AsyncJobWhT) startAsyncJobRunner(ctx context.Context) {
+	asyncWhJob.log.Info("WH-Jobs: Starting async job runner")
+	for {
+		asyncWhJob.log.Info("WH-Jobs: Scanning for waiting async job")
+		select {
+		case <-ctx.Done():
+			asyncWhJob.log.Info("WH-Jobs: Stopping AsyncJobRunner")
+			return
+		case <-time.After(10 * time.Second):
```

Comment:
Can we create a constant for frequency `10 * time.Second`

Reply:
Done